tools/reviewer/job/impl/CITask.java

```diff
@@ -0,0 +1,122 @@
+import com.google.startupos.common.repo.GitRepo;
+import com.google.startupos.common.repo.GitRepoFactory;
+import com.google.startupos.tools.reviewer.ReviewerProtos.CIRequest;
+import com.google.startupos.tools.reviewer.ReviewerProtos.CIResponse;
+import com.google.startupos.tools.reviewer.ReviewerProtos.Repo;
+import com.google.startupos.tools.reviewer.job.tasks.Task;
+
+import javax.inject.Inject;
+import java.io.IOException;
+import java.util.concurrent.locks.ReentrantLock;
+
+public class CITask extends FirestoreTaskBase implements Task {
+  private static FluentLogger log = FluentLogger.forEnclosingClass();
+
+  private final ReentrantLock lock = new ReentrantLock();
+
+  private FileUtils fileUtils;
+  private GitRepoFactory gitRepoFactory;
+
+  private static String CI_REQUESTS_PATH = "/reviewer/ci/requests";
+  private static String CI_RESPONSES_PATH = "/reviewer/ci/responses";
+
+  @Inject
+  public CITask(
+      FileUtils fileUtils,
+      GitRepoFactory gitRepoFactory,
+      FirestoreClientFactory firestoreClientFactory) {
+    this.fileUtils = fileUtils;
+    this.gitRepoFactory = gitRepoFactory;
+    this.firestoreClientFactory = firestoreClientFactory;
+  }
+
+  @Override
+  public Boolean shouldRun() {
+    return !lock.isLocked();
+  }
+
+  @Override
+  public void run() {
+    if (lock.tryLock()) {
+      try {
+        initializeFirestoreClientIfNull();
+
+        log.atInfo().log("Running CI job (Firestore client @ %s)", firestoreClient);
+
+        CIRequest request =
+            (CIRequest) this.firestoreClient.popDocument(CI_REQUESTS_PATH, CIRequest.newBuilder());
+
+        if (request == null) {
+          // there is no request to process
+          return;
+        }
+
+        if (!fileUtils.folderExists("ci")) {
+          fileUtils.mkdirs("ci");
+        }
+
+        CIResponse.Builder responseBuilder = CIResponse.newBuilder().setRequest(request);
+
+        for (CIRequest.Target target : request.getTargetList()) {
+          Repo repo = target.getRepo();
+
+          String repoPath =
+              fileUtils.joinPaths(fileUtils.getCurrentWorkingDirectory(), "ci", repo.getId());
+
+          GitRepo gitRepo = this.gitRepoFactory.create(repoPath);
+
+          if (fileUtils.folderEmptyOrNotExists(repoPath)) {
+            gitRepo.cloneRepo(repo.getUrl(), repoPath);
+          } else {
+            gitRepo.pull();
+          }
+
+          gitRepo.switchBranch(target.getCommitId());
+
+          log.atInfo().log("Running reviewer-ci.sh in %s", repoPath);
+          CommandLine.CommandResult result =
+              CommandLine.runCommandForError(
+                  "/usr/bin/env bash " + fileUtils.joinPaths(repoPath, "reviewer-ci.sh"));
+
+          responseBuilder.addResults(
+              CIResponse.TargetResult.newBuilder()
+                  .setTarget(target)
+                  .setSuccess(result.exitValue == 0)
+                  // TODO: log truncation strategy?
```

Comment:
Let's rephease:
`// TODO: Firestore's limit is 1MB. Truncate log to 950Kb, to leave room for the rest of the message.`
Actually, you can easily do that now, if you can figure out what's the relation between the string length and the encoded proto length (is 1 char = 1 byte or 2 bytes, or something else). If that's hard, we can play it on the safe side and assume it's 2 bytes, and leave a TODO to improve that later.

Reply:
@oferb partially resolved in 5fb8d19b2722c9c38e1eb528d1d9c75c5b40afae